ibc/store/proofs_ics23.go

```diff
@@ -0,0 +1,177 @@
+func createNonMembershipProof(tree *smt.SMT, key []byte) (*ics23.CommitmentProof, error) {
+	proof, err := tree.Prove(key)
+	if err != nil {
+		return nil, coreTypes.ErrCreatingProof(err)
+	}
+
+	return convertSMPToExclusionProof(&proof, key), nil
+}
+
+// convertSMPToExistenceProof converts a SparseMerkleProof to an ics23
+// ExistenceProof to verify membership of an element
+func convertSMPToExistenceProof(proof *smt.SparseMerkleProof, key, value []byte) *ics23.CommitmentProof {
+	path := sha256.Sum256(key)
+	steps := convertSideNodesToSteps(proof.SideNodes, path[:])
+	return &ics23.CommitmentProof{
+		Proof: &ics23.CommitmentProof_Exist{
+			Exist: &ics23.ExistenceProof{
+				Key:   key,
+				Value: value,
+				Leaf:  smtSpec.LeafSpec,
+				Path:  steps,
+			},
+		},
+	}
+}
+
+// convertSMPToExclusionProof converts a SparseMerkleProof to an ics23
+// ExclusionProof to verify non-membership of an element
+func convertSMPToExclusionProof(proof *smt.SparseMerkleProof, key []byte) *ics23.CommitmentProof {
+	path := sha256.Sum256(key)
+	steps := convertSideNodesToSteps(proof.SideNodes, path[:])
+	leaf := &ics23.LeafOp{
+		Hash: ics23.HashOp_SHA256,
+		// Do not re-hash already hashed fields from NonMembershipLeafData
+		PrehashKey:   ics23.HashOp_NO_HASH,
+		PrehashValue: ics23.HashOp_NO_HASH,
+		Length:       ics23.LengthOp_NO_PREFIX,
+		Prefix:       []byte{0},
+	}
+	actualPath := path[:]
+	actualValue := defaultValue
+	if proof.NonMembershipLeafData != nil {
+		actualPath = proof.NonMembershipLeafData[1 : 1+hashSize] // len(prefix): len(prefix) + hashSize
+		actualValue = proof.NonMembershipLeafData[1+hashSize:]
+	}
+	return &ics23.CommitmentProof{
+		Proof: &ics23.CommitmentProof_Exclusion{
+			Exclusion: &ics23.ExclusionProof{
+				Key:             key,
+				ActualPath:      actualPath,
+				ActualValueHash: actualValue,
+				Leaf:            leaf,
+				Path:            steps,
+			},
+		},
+	}
+}
+
+// convertSideNodesToSteps converts the SideNodes field in the SparseMerkleProof
+// into a list of InnerOps for the ics23 CommitmentProof
+func convertSideNodesToSteps(sideNodes [][]byte, path []byte) []*ics23.InnerOp {
+	steps := make([]*ics23.InnerOp, 0, len(sideNodes))
+	for i := 0; i < len(sideNodes); i++ {
+		var prefix, suffix []byte
+		prefix = append(prefix, innerPrefix...)
+		if getPathBit(path, len(sideNodes)-1-i) == left {
+			// path is on the left so sidenode must be on the right
+			suffix = make([]byte, 0, len(sideNodes[i]))
+			suffix = append(suffix, sideNodes[i]...)
+		} else {
+			// path is on the right so sidenode must be on the left
+			prefix = append(prefix, sideNodes[i]...)
+		}
+		op := &ics23.InnerOp{
+			Hash:   ics23.HashOp_SHA256,
+			Prefix: prefix,
+			Suffix: suffix,
+		}
+		steps = append(steps, op)
+	}
+	return steps
+}
+
+// getPathBit takes the hash of a key (the path) and a position (depth) and returns whether at
```

Comment:
See the ref - `Ref: https://github.com/pokt-network/smt/blob/main/utils.go`

Yea, this is definitely non-trivial to implement...

This business logic is VERY closely coupled with our specific SMT implementation. I'd suggest the following:
1. exposing `GetPathBit` in the smt and reusing it here
2. Move the comments you added to the smt
3. Add an `isLeftChild` helper and use it here

Reply:
I have added a techdebt comment here as I have opened [smt#14](https://github.com/pokt-network/smt/issues/14) which will cover this. As I have just release `v0.6.0` for the SMT i think that its best to bundle these small changes into releases to save on constantly updating the `go.mod`. As the functionality will not change until this issue in in progress I think its alright to leave it until then.

LMK wdyt

Reply:
Going to push back on this.

I think https://github.com/pokt-network/smt/issues/14 is a 1+ month research and development project, whereas exposing the `GetPathBit` function is something we can do before EOD.

Reply:
@Olshansk https://github.com/pokt-network/smt/pull/15